apps/emqx/src/emqx_cluster_rpc.erl

```diff
@@ -0,0 +1,322 @@
+%%
+%% Licensed under the Apache License, Version 2.0 (the "License");
+%% you may not use this file except in compliance with the License.
+%% You may obtain a copy of the License at
+%%
+%%     http://www.apache.org/licenses/LICENSE-2.0
+%%
+%% Unless required by applicable law or agreed to in writing, software
+%% distributed under the License is distributed on an "AS IS" BASIS,
+%% WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+%% See the License for the specific language governing permissions and
+%% limitations under the License.
+%%--------------------------------------------------------------------
+-module(emqx_cluster_rpc).
+-behaviour(gen_statem).
+
+%% API
+-export([start_link/2, mnesia/1]).
+-export([multicall/3, query/1, reset/0, status/0]).
+
+-export([init/1, format_status/2, handle_event/4, terminate/3,
+    code_change/4, callback_mode/0]).
+
+-ifdef(TEST).
+-compile(export_all).
+-compile(nowarn_export_all).
+-endif.
+
+-boot_mnesia({mnesia, [boot]}).
+-copy_mnesia({mnesia, [copy]}).
+
+-include("emqx.hrl").
+-include("logger.hrl").
+-define(CLUSTER_MFA, cluster_rpc_mfa).
+-define(CLUSTER_CURSOR, cluster_rpc_cursor).
+
+-rlog_shard({?COMMON_SHARD, ?CLUSTER_MFA}).
+-rlog_shard({?COMMON_SHARD, ?CLUSTER_CURSOR}).
+
+-define(CATCH_UP, catch_up).
+-define(IDLE, idle).
+-define(MFA_HISTORY_LEN, 100).
+
+-record(cluster_rpc_mfa, {tnx_id :: pos_integer(), mfa :: mfa(),
+    created_at :: calendar:datetime(), initiator :: node()}).
+-record(cluster_rpc_cursor, {node :: node(), tnx_id :: pos_integer()}).
+
+%%%===================================================================
+%%% API
+%%%===================================================================
+mnesia(boot) ->
+    ok = ekka_mnesia:create_table(?CLUSTER_MFA, [
+        {type, ordered_set},
+        {disc_copies, [node()]},
+        {local_content, true},
+        {record_name, cluster_rpc_mfa},
+        {attributes, record_info(fields, cluster_rpc_mfa)}]),
+    ok = ekka_mnesia:create_table(?CLUSTER_CURSOR, [
+        {type, set},
+        {disc_copies, [node()]},
+        {local_content, true},
+        {record_name, cluster_rpc_cursor},
+        {attributes, record_info(fields, cluster_rpc_cursor)}]);
+mnesia(copy) ->
+    ok = ekka_mnesia:copy_table(cluster_rpc_mfa, disc_copies),
+    ok = ekka_mnesia:copy_table(cluster_rpc_cursor, disc_copies).
+
+start_link(Node, Name) ->
+    gen_statem:start_link({local, Name}, ?MODULE, [Node], []).
+
+-spec multicall(Module, Function, Args) -> {ok, TnxId} |{error, Reason} when
+    Module :: module(),
+    Function :: atom(),
+    Args :: [term()],
+    TnxId :: pos_integer(),
+    Reason :: term().
+multicall(M, F, A) ->
+    case ekka_rlog:core_nodes() of
+        [] -> {error, "core_nodes is empty"};
+        [Core | _] = CoreNodes ->
+            case lists:member(node(), CoreNodes) of
+                true -> gen_statem:call(?MODULE, {commit, {M, F, A}}, 2 * 60 * 1000);
```

Comment:
`commit` here is perhaps not the best name.
commit usually is an action after some preparation, but the preparation is not done yet before the gen_statem call.

So maybe `initiate` ?